File: src/app/services/order.service.ts
import { Injectable, Inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { APP_CONFIG } from '../../main';

export interface OrderItem {
  id: number;
  qteOrdered: number;
  orderDate: string;
  productId: number;
  userId: number;
  mobile?: string;
  status: string;
  paymentStatus: string;
  paymentMethod?: 'CARD' | 'COD';
  approvedBy?: number;
  approvedAt?: string;
}

export interface CreateOrderRequest {
  productId: number;
  qteOrdered: number;
  mobile: string;
  paymentMethod: 'CARD' | 'COD';
}

@Injectable({ providedIn: 'root' })
export class OrderService {
  private readonly apiUrl: string;

  constructor(
    private readonly http: HttpClient,
    @Inject(APP_CONFIG) private readonly config: any
  ) {
    const base = (this.config?.apiUrl || '').replace(/\/$/, '');
    this.apiUrl = base.replace(/\/api$/, '') + '/api/orders';
  }

  getMyOrders(): Observable<OrderItem[]> {
    return this.http.get<OrderItem[]>(`${this.apiUrl}/my-orders`);
  }

  // ← ADD THIS METHOD
  createOrder(payload: CreateOrderRequest): Observable<OrderItem> {
    return this.http.post<OrderItem>(`${this.apiUrl}/place-order`, payload);
  }

  getAllOrders(): Observable<OrderItem[]> {
    const url = `${this.apiUrl}/all`;
    return this.http.get<OrderItem[]>(url);
  }

  approveOrder(orderId: number, force: boolean = false): Observable<OrderItem> {
    const url = `${this.apiUrl}/${orderId}/approve${force ? '?force=true' : ''}`;
    return this.http.put<OrderItem>(url, {});
  }

  // Admin declines or deletes (cancel) an order
  // Admin decline -> calls admin decline endpoint
  declineOrder(orderId: number): Observable<OrderItem> {
    const url = `${this.apiUrl}/${orderId}/decline`;
    return this.http.put<OrderItem>(url, {});
  }

  // Admin permanent delete
  deleteOrder(orderId: number): Observable<void> {
    const url = `${this.apiUrl}/${orderId}/admin`;
    return this.http.delete<void>(url);
  }

  // Mark payment status (used by admin to mark COD as paid when delivered)
  markPaymentPaid(orderId: number) {
    const url = `${this.apiUrl}/${orderId}/payment-status?paymentStatus=PAID`;
    return this.http.put<OrderItem>(url, {});
  }

  markAsDelivered(orderId: number): Observable<OrderItem> {
  const url = `${this.apiUrl}/${orderId}/delivered`;
  return this.http.put<OrderItem>(url, {});
}
}

File: lib/services/order_service.dart
import 'package:flutter/material.dart';
import '../models/order.dart';
import 'api_service.dart';

class OrderService extends ChangeNotifier {
  final ApiService _apiService = ApiService();

  Future<List<OrderItem>> getMyOrders() async {
    try {
      final response = await _apiService.get('/orders/my-orders');
      // response is already decoded JSON (List<dynamic>)
      final List<dynamic> data = response as List<dynamic>;
      return data.map((json) => OrderItem.fromJson(json)).toList();
    } catch (e) {
      print('Error fetching orders: $e');
      rethrow;
    }
  }

  // Admin: get all orders
  Future<List<OrderItem>> getAllOrders() async {
    final response = await _apiService.get('/orders/all');
    final List<dynamic> data = response as List<dynamic>;
    return data.map((json) => OrderItem.fromJson(json)).toList();
  }

  Future<OrderItem> approveOrder(int id, {bool force = false}) async {
    final q = force ? '?force=true' : '';
    final response = await _apiService.put('/orders/$id/approve$q', {});
    return OrderItem.fromJson(Map<String, dynamic>.from(response as Map));
  }

  Future<OrderItem> markAsDelivered(int id) async {
    final response = await _apiService.put('/orders/$id/delivered', {});
    return OrderItem.fromJson(Map<String, dynamic>.from(response as Map));
  }

  Future<OrderItem> declineOrder(int id) async {
    final response = await _apiService.put('/orders/$id/decline', {});
    return OrderItem.fromJson(Map<String, dynamic>.from(response as Map));
  }

  Future<void> deleteOrder(int id) async {
    await _apiService.delete('/orders/$id/admin');
  }
}
